File: JavascriptAlgorithm/zerotrac/1500_1600/tupleSameProduct_1726.js
var tupleSameProduct = function (nums) {
    const n = nums.length;
    const map = new Map();
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            let mul = nums[i] * nums[j];
            map.set(mul, (map.get(mul) || 0) + 1);
        }
    }
    /*
    map 计数大于 1，则说明有两对数满足条件
    假设计数为2： 2*6 == 4*3 则贡献的数目为：2*2*2 -> [2*6 6*2] [3*4  4*3] 
    同时可以交换位置 2*6放置在左侧
    计数 > 2 时选择排列组合 （4，2）
    */
    let res = 0;
    for (let v of map.values()) {
        if (v > 1) res += (v * (v - 1) / 2 * 8);
    }
    return res;
};

File: Clion/ZeroTrac/1500_1600/tupleSameProduct_1726.cpp
//
// Created by TzashiNorpu on 7/11/2022.
//
#include "header.h"

using namespace ZeroTrac;

bool valid(vector<int> &nums) {
    int a = nums[0];
    int b = nums[1];
    int c = nums[2];
    int d = nums[3];
    return (a * b == c * d) && (a != b && b != c && c != d);
}

void tupleSameProduct_1726::helper(vector<int> &nums, vector<int> &choose, vector<bool> chooseIdx) {
    if (choose.size() == 4) {
        if (valid(choose)) {
            res++;
        }
        return;
    }
    for (int i = 0; i < nums.size(); i++) {
        if (!chooseIdx[i]) {
            choose.push_back(nums[i]);
            chooseIdx[i] = true;
            helper(nums, choose, chooseIdx);
            chooseIdx[i] = false;
            choose.pop_back();
        }
    }
}

int tupleSameProduct_1726::tupleSameProduct(vector<int> &nums) {
    vector<int> choose;
    vector<bool> chooseIdx(nums.size(), false);
    helper(nums, choose, chooseIdx);
    return res;
}


